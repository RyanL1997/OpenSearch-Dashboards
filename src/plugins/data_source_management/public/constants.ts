/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { i18n } from '@osd/i18n';
import { DirectQueryDatasourceType } from './types';

export const QUERY_RESTRICTED = 'query-restricted';
export const QUERY_ALL = 'query-all';
export const LOCAL_CLUSTER = 'local_cluster';

export const DatasourceTypeToDisplayName: { [key in DirectQueryDatasourceType]: string } = {
  PROMETHEUS: 'Prometheus',
  S3GLUE: 'Amazon S3',
};

export const PROMETHEUS_URL = 'Prometheus';
export const AMAZON_S3_URL = 'AmazonS3AWSGlue';
export const OPENSEARCH_URL = 'OpenSearch';

export const UrlToDatasourceType: { [key: string]: DirectQueryDatasourceType } = {
  [PROMETHEUS_URL]: 'PROMETHEUS',
  [AMAZON_S3_URL]: 'S3GLUE',
};

export const DATACONNECTIONS_BASE = '/api/directquery/dataconnections';
export const SECURITY_ROLES = '/api/v1/configuration/roles';
export const EDIT = '/edit';
export const DATACONNECTIONS_UPDATE_STATUS = '/status';
export const INTEGRATIONS_BASE = '/api/integrations';
export const observabilityMetricsID = 'observability-metrics';

// Module for handling EMR states for Dashboards Progress Bar. All of these except "initial" are
// directly from the EMR job run states. "ord" is used to approximate progress (eyeballed relative
// stage times), and "terminal" indicates whether a job is in progress at all.
export const EMR_STATES = new Map<string, { ord: number; terminal: boolean }>([
  ['submitted', { ord: 0, terminal: false }],
  ['queued', { ord: 10, terminal: false }],
  ['pending', { ord: 20, terminal: false }],
  ['scheduled', { ord: 30, terminal: false }],
  ['running', { ord: 70, terminal: false }],
  ['cancelling', { ord: 90, terminal: false }],
  ['success', { ord: 100, terminal: true }],
  ['failed', { ord: 100, terminal: true }],
  ['canceled', { ord: 100, terminal: true }],
  // The "null state" for a initial page load, which components conditionally use on load.
  ['initial', { ord: 100, terminal: true }],
]);

export const MAX_ORD = 100;

export function intervalAsMinutes(interval: number): string {
  const minutes = Math.floor(interval / 60000);
  return minutes === 1
    ? i18n.translate('dataSourcesManagement.directQuerySync.intervalAsMinutes.oneMinute', {
        defaultMessage: '1 minute',
      })
    : i18n.translate('dataSourcesManagement.directQuerySync.intervalAsMinutes.multipleMinutes', {
        defaultMessage: '{minutes} minutes',
        values: { minutes },
      });
}
