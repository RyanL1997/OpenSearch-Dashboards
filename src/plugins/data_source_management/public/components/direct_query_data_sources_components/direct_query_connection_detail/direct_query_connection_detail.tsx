/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useEffect, useState } from 'react';
import {
  EuiPage,
  EuiPageBody,
  EuiPageHeader,
  EuiPageHeaderSection,
  EuiFlexGroup,
  EuiFlexItem,
  EuiTitle,
  EuiPanel,
  EuiSpacer,
  EuiText,
  EuiTabbedContent,
} from '@elastic/eui';
import {
  HttpStart,
  IUiSettingsClient,
  NotificationsStart,
  SavedObjectsStart,
} from 'opensearch-dashboards/public';
import { useLocation, useParams } from 'react-router-dom';
import { DATACONNECTIONS_BASE } from '../../../constants';
import { DirectQueryDatasourceDetails, PrometheusProperties } from '../../../types';
import { NoAccess } from './no_access_page';
import { InactiveDataConnectionCallout } from './inactive_data_connection_callout';
import { AccessControlTab } from './access_control_tab';
import { getManageDirectQueryDataSourceBreadcrumbs } from '../../breadcrumbs';
import { useLoadAccelerationsToCache } from '../../../../framework/catlog_cache/cache_loader';
import { AccelerationTable } from '../direct_query_acceleration_management/acceleration_table';

interface DirectQueryDataConnectionDetailProps {
  featureFlagStatus: boolean;
  http: HttpStart;
  notifications: NotificationsStart;
  savedObjects: SavedObjectsStart;
  uiSettings: IUiSettingsClient;
  setBreadcrumbs: (breadcrumbs: any) => void;
}

export const DirectQueryDataConnectionDetail: React.FC<DirectQueryDataConnectionDetailProps> = ({
  featureFlagStatus,
  http,
  notifications,
  savedObjects,
  uiSettings,
  setBreadcrumbs,
}) => {
  const { dataSourceName } = useParams<{ dataSourceName: string }>();
  const { search } = useLocation();
  const queryParams = new URLSearchParams(search);
  const dataSourceMDSId = queryParams.get('dataSourceMDSId');
  const [datasourceDetails, setDatasourceDetails] = useState<DirectQueryDatasourceDetails>({
    allowedRoles: [],
    name: '',
    description: '',
    connector: 'PROMETHEUS',
    properties: { 'prometheus.uri': 'placeholder' },
    status: 'ACTIVE',
  });
  const [hasAccess, setHasAccess] = useState(true);

  // Cache loader hook
  const {
    loadStatus: accelerationsLoadStatus,
    startLoading: startLoadingAccelerations,
  } = useLoadAccelerationsToCache();

  const cacheLoadingHooks = {
    accelerationsLoadStatus,
    startLoadingAccelerations,
  };

  const fetchSelectedDatasource = () => {
    http
      .get(`${DATACONNECTIONS_BASE}/${dataSourceName}`)
      .then((data) => {
        setDatasourceDetails({
          allowedRoles: data.allowedRoles,
          description: data.description,
          name: data.name,
          connector: data.connector,
          properties: data.properties,
          status: data.status,
        });
      })
      .catch((_err) => {
        setHasAccess(false);
      });
  };

  useEffect(() => {
    setBreadcrumbs(getManageDirectQueryDataSourceBreadcrumbs(dataSourceName));
    fetchSelectedDatasource();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [setBreadcrumbs, http]);

  const S3DatasourceOverview = () => (
    <EuiPanel>
      <EuiFlexGroup>
        <EuiFlexItem>
          <EuiFlexGroup direction="column">
            <EuiFlexItem grow={false}>
              <EuiText className="overview-title">Description</EuiText>
              <EuiText size="s" className="overview-content">
                {datasourceDetails.description || '-'}
              </EuiText>
            </EuiFlexItem>
          </EuiFlexGroup>
        </EuiFlexItem>
        <EuiFlexItem>
          <EuiFlexGroup direction="column">
            <EuiFlexItem grow={false}>
              <EuiText className="overview-title">Query Access</EuiText>
              <EuiText size="s" className="overview-content">
                {datasourceDetails.allowedRoles.length > 0
                  ? `Restricted to ${datasourceDetails.allowedRoles.join(', ')}`
                  : 'Admin only'}
              </EuiText>
            </EuiFlexItem>
          </EuiFlexGroup>
        </EuiFlexItem>
      </EuiFlexGroup>
      <EuiSpacer />
    </EuiPanel>
  );

  const PrometheusDatasourceOverview = () => (
    <EuiPanel>
      <EuiFlexGroup>
        <EuiFlexItem>
          <EuiFlexGroup direction="column">
            <EuiFlexItem grow={false}>
              <EuiText className="overview-title">Connection title</EuiText>
              <EuiText size="s" className="overview-content">
                {datasourceDetails.name || '-'}
              </EuiText>
            </EuiFlexItem>
            <EuiFlexItem grow={false}>
              <EuiText className="overview-title">Data source description</EuiText>
              <EuiText size="s" className="overview-content">
                {datasourceDetails.description || '-'}
              </EuiText>
            </EuiFlexItem>
          </EuiFlexGroup>
        </EuiFlexItem>
        <EuiFlexItem>
          <EuiFlexGroup direction="column">
            <EuiFlexItem grow={false}>
              <EuiText className="overview-title">Prometheus URI</EuiText>
              <EuiText size="s" className="overview-content">
                {(datasourceDetails.properties as PrometheusProperties)['prometheus.uri'] || '-'}
              </EuiText>
            </EuiFlexItem>
          </EuiFlexGroup>
        </EuiFlexItem>
      </EuiFlexGroup>
      <EuiSpacer />
    </EuiPanel>
  );

  const DatasourceOverview = () => {
    switch (datasourceDetails.connector) {
      case 'S3GLUE':
        return <S3DatasourceOverview />;
      case 'PROMETHEUS':
        return <PrometheusDatasourceOverview />;
      default:
        return null;
    }
  };

  if (!hasAccess) {
    return <NoAccess />;
  }

  const genericTabs = [
    {
      id: 'access_control',
      name: 'Access control',
      disabled: false,
      content: (
        <AccessControlTab
          dataConnection={dataSourceName}
          connector={datasourceDetails.connector}
          properties={datasourceDetails.properties}
          allowedRoles={datasourceDetails.allowedRoles}
          key={JSON.stringify(datasourceDetails.allowedRoles)}
        />
      ),
    },
  ];

  const conditionalTabs =
    datasourceDetails.connector === 'S3GLUE'
      ? [
          {
            id: 'acceleration_table',
            name: 'Accelerations',
            disabled: false,
            content: (
              <AccelerationTable
                dataSourceName={dataSourceName}
                cacheLoadingHooks={cacheLoadingHooks}
                http={http}
                notifications={notifications}
              />
            ),
          },
        ]
      : [];

  const tabs = [...conditionalTabs, ...genericTabs];

  return (
    <EuiPage>
      <EuiPageBody>
        <EuiPageHeader style={{ justifyContent: 'spaceBetween' }}>
          <EuiPageHeaderSection style={{ width: '100%', justifyContent: 'space-between' }}>
            <EuiFlexGroup>
              <EuiFlexItem grow={false}>
                <EuiTitle data-test-subj="datasourceTitle" size="l">
                  <h1>{datasourceDetails.name}</h1>
                </EuiTitle>
              </EuiFlexItem>
            </EuiFlexGroup>
          </EuiPageHeaderSection>
        </EuiPageHeader>
        <DatasourceOverview />
        <EuiSpacer />
        {datasourceDetails.status !== 'ACTIVE' ? (
          <InactiveDataConnectionCallout
            datasourceDetails={datasourceDetails}
            fetchSelectedDatasource={fetchSelectedDatasource}
          />
        ) : (
          <>
            {/* <EuiAccordion
              id="queryOrAccelerateAccordion"
              buttonContent={'Get started'}
              initialIsOpen={true}
              paddingSize="m"
            >
              <QueryOrAccelerateData />
            </EuiAccordion> */}
            <EuiTabbedContent tabs={tabs} />
          </>
        )}

        <EuiSpacer />
      </EuiPageBody>
    </EuiPage>
  );
};
