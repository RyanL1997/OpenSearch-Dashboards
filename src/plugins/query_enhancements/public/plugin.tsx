/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { i18n } from '@osd/i18n';
import { CoreSetup, CoreStart, Plugin, PluginInitializerContext } from '../../../core/public';
import { ConfigSchema } from '../common/config';
import { setData, setStorage } from './services';
import { createQueryAssistExtension } from './query_assist';
import { PPLSearchInterceptor, SQLSearchInterceptor } from './search';
import {
  QueryEnhancementsPluginSetup,
  QueryEnhancementsPluginSetupDependencies,
  QueryEnhancementsPluginStart,
  QueryEnhancementsPluginStartDependencies,
} from './types';
import { LanguageConfig, Query } from '../../data/public';
import { s3TypeConfig } from './datasets';
import { createEditor, DefaultInput, SingleLineInput } from '../../data/public';
import { DataStorage } from '../../data/common';
import { pplLanguageReference, sqlLanguageReference } from './query_editor_extensions';

export class QueryEnhancementsPlugin
  implements
    Plugin<
      QueryEnhancementsPluginSetup,
      QueryEnhancementsPluginStart,
      QueryEnhancementsPluginSetupDependencies,
      QueryEnhancementsPluginStartDependencies
    > {
  private readonly storage: DataStorage;
  private readonly config: ConfigSchema;

  constructor(initializerContext: PluginInitializerContext) {
    this.config = initializerContext.config.get<ConfigSchema>();
    this.storage = new DataStorage(window.localStorage, 'discover.queryAssist.');
  }

  public setup(
    core: CoreSetup<QueryEnhancementsPluginStartDependencies>,
    { data, usageCollection }: QueryEnhancementsPluginSetupDependencies
  ): QueryEnhancementsPluginSetup {
    const { queryString } = data.query;

    // Initialize local storage keys when the plugin loads, if they don't exist
    if (!window.localStorage.getItem('hasSeenSQLInfoBox')) {
      window.localStorage.setItem('hasSeenSQLInfoBox', 'false');
    }
    if (!window.localStorage.getItem('hasSeenPPLInfoBox')) {
      window.localStorage.setItem('hasSeenPPLInfoBox', 'false');
    }

    // Initialize `selectedLanguage` as undefined to wait for correct language updates
    let selectedLanguage;

    // Subscribe to updates to track language changes
    queryString.getUpdates$().subscribe((query) => {
      selectedLanguage = query.language;

      // Re-render the controls when `selectedLanguage` is defined
      const pplControls = [pplLanguageReference(selectedLanguage)];
      const sqlControls = [sqlLanguageReference(selectedLanguage)];

      const enhancedPPLQueryEditor = createEditor(SingleLineInput, null, pplControls, DefaultInput);
      const enhancedSQLQueryEditor = createEditor(SingleLineInput, null, sqlControls, DefaultInput);

      // Define and register PPL language configuration
      const pplLanguageConfig: LanguageConfig = {
        id: 'PPL',
        title: 'PPL',
        search: new PPLSearchInterceptor({
          toasts: core.notifications.toasts,
          http: core.http,
          uiSettings: core.uiSettings,
          startServices: core.getStartServices(),
          usageCollector: data.search.usageCollector,
        }),
        getQueryString: (currentQuery: Query) => `source = ${currentQuery.dataset?.title}`,
        fields: { filterable: false, visualizable: false },
        docLink: {
          title: i18n.translate('queryEnhancements.pplLanguage.docLink', {
            defaultMessage: 'PPL documentation',
          }),
          url: 'https://opensearch.org/docs/latest/search-plugins/sql/ppl/syntax/',
        },
        showDocLinks: false,
        editor: enhancedPPLQueryEditor,
        editorSupportedAppNames: ['discover'],
        supportedAppNames: ['discover', 'data-explorer'],
      };
      queryString.getLanguageService().registerLanguage(pplLanguageConfig);

      // Define and register SQL language configuration
      const sqlLanguageConfig: LanguageConfig = {
        id: 'SQL',
        title: 'OpenSearch SQL',
        search: new SQLSearchInterceptor({
          toasts: core.notifications.toasts,
          http: core.http,
          uiSettings: core.uiSettings,
          startServices: core.getStartServices(),
          usageCollector: data.search.usageCollector,
        }),
        getQueryString: (currentQuery: Query) =>
          `SELECT * FROM ${currentQuery.dataset?.title} LIMIT 10`,
        fields: { filterable: false, visualizable: false },
        docLink: {
          title: i18n.translate('queryEnhancements.sqlLanguage.docLink', {
            defaultMessage: 'SQL documentation',
          }),
          url: 'https://opensearch.org/docs/latest/search-plugins/sql/sql/basic/',
        },
        showDocLinks: false,
        editor: enhancedSQLQueryEditor,
        editorSupportedAppNames: ['discover'],
        supportedAppNames: ['discover', 'data-explorer'],
        hideDatePicker: true,
        sampleQueries: [
          {
            title: i18n.translate('queryEnhancements.sqlLanguage.sampleQuery.titleContainsWind', {
              defaultMessage: 'The title field contains the word wind.',
            }),
            query: `SELECT * FROM your_table WHERE title LIKE '%wind%'`,
          },
          {
            title: i18n.translate(
              'queryEnhancements.sqlLanguage.sampleQuery.titleContainsWindOrWindy',
              {
                defaultMessage: 'The title field contains the word wind or the word windy.',
              }
            ),
            query: `SELECT * FROM your_table WHERE title LIKE '%wind%' OR title LIKE '%windy%';`,
          },
          {
            title: i18n.translate(
              'queryEnhancements.sqlLanguage.sampleQuery.titleContainsPhraseWindRises',
              {
                defaultMessage: 'The title field contains the phrase wind rises.',
              }
            ),
            query: `SELECT * FROM your_table WHERE title LIKE '%wind rises%'`,
          },
          {
            title: i18n.translate(
              'queryEnhancements.sqlLanguage.sampleQuery.titleExactMatchWindRises',
              {
                defaultMessage: 'The title.keyword field exactly matches The wind rises.',
              }
            ),
            query: `SELECT * FROM your_table WHERE title = 'The wind rises'`,
          },
          {
            title: i18n.translate(
              'queryEnhancements.sqlLanguage.sampleQuery.titleFieldsContainWind',
              {
                defaultMessage:
                  'Any field that starts with title (for example, title and title.keyword) contains the word wind',
              }
            ),
            query: `SELECT * FROM your_table WHERE title LIKE '%wind%' OR title = 'wind'`,
          },
          {
            title: i18n.translate(
              'queryEnhancements.sqlLanguage.sampleQuery.descriptionFieldExists',
              {
                defaultMessage: 'Documents in which the field description exists.',
              }
            ),
            query: `SELECT * FROM your_table WHERE description IS NOT NULL AND description != '';`,
          },
        ],
      };
      queryString.getLanguageService().registerLanguage(sqlLanguageConfig);
    });

    data.__enhance({
      editor: {
        queryEditorExtension: createQueryAssistExtension(
          core,
          data,
          this.config.queryAssist,
          usageCollection
        ),
      },
    });

    queryString.getDatasetService().registerType(s3TypeConfig);

    return {};
  }

  public start(
    core: CoreStart,
    deps: QueryEnhancementsPluginStartDependencies
  ): QueryEnhancementsPluginStart {
    setStorage(this.storage);
    setData(deps.data);
    return {};
  }

  public stop() {}
}
