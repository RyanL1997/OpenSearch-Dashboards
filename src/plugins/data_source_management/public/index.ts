/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { DataSourceManagementPlugin } from './plugin';

// This exports static code and TypeScript types,
// as well as, OpenSearch Dashboards Platform `plugin()` initializer.
export function plugin() {
  return new DataSourceManagementPlugin();
}

export { DataSourceManagementPluginStart, DirectQueryDatasourceDetails } from './types';
export { DataSourceSelector, DataSourceOption } from './components/data_source_selector';
export { DataSourceMenu } from './components/data_source_menu';
export { DataSourceManagementPlugin, DataSourceManagementPluginSetup } from './plugin';
export {
  DataSourceSelectableConfig,
  DataSourceComponentType,
  DataSourceAggregatedViewConfig,
  DataSourceViewConfig,
  DataSourceMenuProps,
  DataSourceMultiSelectableConfig,
  createDataSourceMenu,
} from './components/data_source_menu';
export { DataSourceSelectionService } from './service/data_source_selection_service';
export { getDefaultDataSourceId, getDefaultDataSourceId$ } from './components/utils';
export { DATACONNECTIONS_BASE, DatasourceTypeToDisplayName } from './constants';

// Export framework utilities
export { usePolling, UsePolling, PollingConfigurations } from '../framework/utils/use_polling';
export { SQLService } from '../framework/requests/sql';
export { useDirectQuery } from '../framework/hooks/direct_query_hook';
export { DirectQueryRequest, DirectQueryLoadingStatus } from '../framework/types';
export { getAsyncSessionId, setAsyncSessionId } from '../framework/utils/query_session_utils';
export { formatError } from '../framework/utils/shared';
export { ASYNC_POLLING_INTERVAL } from '../framework/constants';
