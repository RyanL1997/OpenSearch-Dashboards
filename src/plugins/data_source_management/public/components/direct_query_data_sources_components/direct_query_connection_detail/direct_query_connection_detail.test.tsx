/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import { MemoryRouter, Route } from 'react-router-dom';
import { DirectQueryDataConnectionDetail } from './direct_query_connection_detail';
import { ApplicationStart, HttpStart, NotificationsStart } from 'opensearch-dashboards/public';

// Mock the imported modules and components
jest.mock('../../../constants', () => ({
  DATACONNECTIONS_BASE: '/api/dataconnections',
}));

jest.mock('./utils/no_access_page', () => ({
  NoAccess: () => <div>No Access</div>,
}));

jest.mock('./utils/inactive_data_connection_callout', () => ({
  InactiveDataConnectionCallout: () => <div>Inactive Data Connection</div>,
}));

jest.mock('./access_control_tab', () => ({
  AccessControlTab: () => <div>Access Control Tab</div>,
}));

jest.mock('../../breadcrumbs', () => ({
  getManageDirectQueryDataSourceBreadcrumbs: () => [{ text: 'Breadcrumb' }],
}));

jest.mock('../../../../framework/catalog_cache/cache_loader', () => ({
  useLoadAccelerationsToCache: jest.fn(() => ({
    loadStatus: 'IDLE',
    startLoading: jest.fn(),
  })),
  useLoadDatabasesToCache: jest.fn(() => ({
    loadStatus: 'IDLE',
    startLoading: jest.fn(),
  })),
  useLoadTablesToCache: jest.fn(() => ({
    loadStatus: 'IDLE',
    startLoading: jest.fn(),
  })),
}));

jest.mock('../direct_query_acceleration_management/acceleration_table', () => ({
  AccelerationTable: () => <div>Acceleration Table</div>,
}));

jest.mock('../../../plugin', () => ({
  getRenderCreateAccelerationFlyout: () => jest.fn(),
}));

jest.mock('../direct_query_associated_object_management/associated_objects_tab', () => ({
  AssociatedObjectsTab: () => <div>Associated Objects Tab</div>,
}));

jest.mock(
  '../direct_query_associated_object_management/utils/associated_objects_tab_utils',
  () => ({
    redirectToExplorerS3: jest.fn(),
  })
);

const renderComponent = ({
  featureFlagStatus = false,
  http = {},
  notifications = {},
  application = {},
  setBreadcrumbs = jest.fn(),
}) => {
  return render(
    <MemoryRouter initialEntries={['/dataconnections/test']}>
      <Route path="/dataconnections/:dataSourceName">
        <DirectQueryDataConnectionDetail
          featureFlagStatus={featureFlagStatus}
          http={http as HttpStart}
          notifications={notifications as NotificationsStart}
          application={application as ApplicationStart}
          setBreadcrumbs={setBreadcrumbs}
        />
      </Route>
    </MemoryRouter>
  );
};

describe('DirectQueryDataConnectionDetail', () => {
  test('renders without crashing', async () => {
    const mockHttp = {
      get: jest.fn().mockResolvedValue({
        allowedRoles: ['role1'],
        description: 'Test description',
        name: 'Test datasource',
        connector: 'PROMETHEUS',
        properties: { 'prometheus.uri': 'placeholder' },
        status: 'ACTIVE',
      }),
    };

    renderComponent({ http: mockHttp });

    await waitFor(() => {
      const titleElement = screen.getByTestId('datasourceTitle');
      expect(titleElement).toBeInTheDocument();
      expect(titleElement).toHaveTextContent('Test datasource');
    });
  });

  test('shows no access message if there is no access', async () => {
    const mockHttp = {
      get: jest.fn().mockRejectedValue(new Error('No access')),
    };

    renderComponent({ http: mockHttp });

    await waitFor(() => {
      expect(screen.getByText('No Access')).toBeInTheDocument();
    });
  });

  test('renders InactiveDataConnectionCallout when datasource is not active', async () => {
    const mockHttp = {
      get: jest.fn().mockResolvedValue({
        allowedRoles: ['role1'],
        description: 'Test description',
        name: 'Test datasource',
        connector: 'PROMETHEUS',
        properties: { 'prometheus.uri': 'placeholder' },
        status: 'INACTIVE',
      }),
    };

    renderComponent({ http: mockHttp });

    await waitFor(() => {
      expect(screen.getByText('Inactive Data Connection')).toBeInTheDocument();
    });
  });

  test('renders AccessControlTab when the tab is clicked', async () => {
    const mockHttp = {
      get: jest.fn().mockResolvedValue({
        allowedRoles: ['role1'],
        description: 'Test description',
        name: 'Test datasource',
        connector: 'PROMETHEUS',
        properties: { 'prometheus.uri': 'placeholder' },
        status: 'ACTIVE',
      }),
    };

    renderComponent({ http: mockHttp });

    await waitFor(() => {
      const titleElement = screen.getByTestId('datasourceTitle');
      expect(titleElement).toBeInTheDocument();
      expect(titleElement).toHaveTextContent('Test datasource');
    });

    // Click the Access Control Tab
    screen.getByText('Access control').click();

    await waitFor(() => {
      expect(screen.getByText('Access Control Tab')).toBeInTheDocument();
    });
  });

  test('matches snapshot', async () => {
    const mockHttp = {
      get: jest.fn().mockResolvedValue({
        allowedRoles: ['role1'],
        description: 'Test description',
        name: 'Test datasource',
        connector: 'PROMETHEUS',
        properties: { 'prometheus.uri': 'placeholder' },
        status: 'ACTIVE',
      }),
    };

    const { asFragment } = renderComponent({ http: mockHttp });

    await waitFor(() => {
      const titleElement = screen.getByTestId('datasourceTitle');
      expect(titleElement).toBeInTheDocument();
      expect(titleElement).toHaveTextContent('Test datasource');
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
