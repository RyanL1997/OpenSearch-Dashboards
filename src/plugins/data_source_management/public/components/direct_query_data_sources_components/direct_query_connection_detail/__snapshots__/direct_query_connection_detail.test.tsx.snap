// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DirectQueryDataConnectionDetail matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="euiPage euiPage--paddingMedium euiPage--grow"
  >
    <div
      class="euiPageBody euiPageBody--borderRadiusNone"
    >
      <header
        class="euiPageHeader euiPageHeader--responsive euiPageHeader--center"
        style="justify-content: spaceBetween;"
      >
        <div
          class="euiPageHeaderSection"
          style="width: 100%; justify-content: space-between;"
        >
          <div
            class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
          >
            <div
              class="euiFlexItem euiFlexItem--flexGrowZero"
            >
              <h1
                class="euiTitle euiTitle--large"
                data-test-subj="datasourceTitle"
              >
                Test datasource
              </h1>
            </div>
          </div>
        </div>
      </header>
      <div
        class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
          >
            <div
              class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionColumn euiFlexGroup--responsive"
            >
              <div
                class="euiFlexItem euiFlexItem--flexGrowZero"
              >
                <div
                  class="euiText euiText--medium overview-title"
                >
                  Connection title
                </div>
                <div
                  class="euiText euiText--small overview-content"
                >
                  Test datasource
                </div>
              </div>
              <div
                class="euiFlexItem euiFlexItem--flexGrowZero"
              >
                <div
                  class="euiText euiText--medium overview-title"
                >
                  Data source description
                </div>
                <div
                  class="euiText euiText--small overview-content"
                >
                  Test description
                </div>
              </div>
            </div>
          </div>
          <div
            class="euiFlexItem"
          >
            <div
              class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionColumn euiFlexGroup--responsive"
            >
              <div
                class="euiFlexItem euiFlexItem--flexGrowZero"
              >
                <div
                  class="euiText euiText--medium overview-title"
                >
                  Prometheus URI
                </div>
                <div
                  class="euiText euiText--small overview-content"
                >
                  placeholder
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="euiSpacer euiSpacer--l"
        />
      </div>
      <div
        class="euiSpacer euiSpacer--l"
      />
      <div
        class="euiAccordion euiAccordion-isOpen"
      >
        <div
          class="euiAccordion__triggerWrapper"
        >
          <button
            aria-controls="queryOrAccelerateAccordion"
            aria-expanded="true"
            class="euiAccordion__button"
            id="generated-id"
            type="button"
          >
            <span
              class="euiAccordion__iconWrapper"
            >
              <span
                class="euiAccordion__icon euiAccordion__icon-isOpen"
                data-euiicon-type="arrowRight"
              />
            </span>
            <span
              class="euiIEFlexWrapFix"
            >
              Get started
            </span>
          </button>
        </div>
        <div
          aria-labelledby="generated-id"
          class="euiAccordion__childWrapper"
          id="queryOrAccelerateAccordion"
          role="region"
          tabindex="-1"
        >
          <div>
            <div
              class="euiAccordion__padding--m"
            >
              <div
                class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
              >
                <div
                  class="euiFlexItem"
                >
                  <div
                    class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow euiPanel--isClickable euiCard euiCard--centerAligned euiCard--isClickable euiCard--hasIcon"
                  >
                    <div
                      class="euiCard__top"
                    >
                      <span
                        class="euiCard__icon"
                        data-euiicon-type="discoverApp"
                      />
                    </div>
                    <div
                      class="euiCard__content"
                    >
                      <span
                        class="euiTitle euiTitle--small euiCard__title"
                        id="generated-idTitle"
                      >
                        <button
                          aria-describedby=" generated-idDescription"
                          class="euiCard__titleButton"
                        >
                          Query data
                        </button>
                      </span>
                      <div
                        class="euiText euiText--small euiCard__description"
                        id="generated-idDescription"
                      >
                        <p>
                          Query your data in Metrics Analytics.
                        </p>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          class="euiTabs"
          role="tablist"
        >
          <button
            aria-controls="generated-id"
            aria-selected="true"
            class="euiTab euiTab-isSelected"
            id="access_control"
            role="tab"
            type="button"
          >
            <span
              class="euiTab__content"
            >
              Access control
            </span>
          </button>
        </div>
        <div
          aria-labelledby="access_control"
          id="generated-id"
          role="tabpanel"
        >
          <div>
            Access Control Tab
          </div>
        </div>
      </div>
      <div
        class="euiSpacer euiSpacer--l"
      />
    </div>
  </div>
</DocumentFragment>
`;
